### DISCUSSION TITLE
Adapt app behavior when power saving mode is enabled

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3251] [Ideas]
My phone's battery is old and terrible, so I'm quite sensitive to battery-draining features. StreetComplete in particular is hard on the battery, because it requires turning up the brightness to see the phone outside, uses gps, and frequently accesses the network, at least when auto-upload is enabled. I get a little over an hour of active mapping on a full charge.

The brightness is unsolvable, it's just needed to use the phone outside. Network access has an easy solution (turn off auto-upload and download data in advance). GPS may have room for improvement.

Now that we've dropped support for api<21, we can able to use [`PowerManager#isPowerSaveMode()`](https://developer.android.com/reference/android/os/PowerManager#isPowerSaveMode()) to check whether power saving is enabled. In that case, perhaps we could lower the frequency of gps updates (and maybe *slightly* decrease the accuracy?).

**Comment by westnordost** [comment]
> I get a little over an hour of active mapping on a full charge.

Whoa. I think you should get a replacement battery to be honest.

> PowerManager#isPowerSaveMode() 

Okay, ... I am not sure if this is going to have any effect on the battery though.

So in general: Sure, we can think about looking for features to turn off in power saver mode. But:
- it must not look like a bug if that feature "does not work" then.
- it must be researched and tested what exactly has an effect on the battery

If you would like to dive into that, I'd suggest to first look at disabling the compass (sensor).

**Reply by smichel17** [reply]
> I think you should get a replacement battery to be honest.

If only it were that simple…

The Galaxy S6 battery is quite difficult to replace; despite having some experience with laptop repair, I'm a little nervous to do this one myself, due to the risk of damaging the back plate (glass), screen, and motherboard. Plus the need to replace the back plate adhesive correctly or else lose some water-resistance. See:

- Guide: https://www.ifixit.com/Guide/Samsung+Galaxy+S6+Battery+Replacement/44648
- Video: https://www.youtube.com/watch?v=mfV4CK6KIek

My current plan is to use the S6 until it doesn't work any more. Hopefully by that time the Librem 5 will be ready enough to replace it. If not, I'll replace it with some cheap used phone and do the same thing; I don't plan to buy an Android flagship. So it's also not worth paying someone to replace the battery; at that point I might as well get a new phone.

I'll probably attempt a self-repair when the battery is *really* preventing use of the phone, but right now, while it's bad, it's not actually *that* bad. It lasts for a full day of normal use, which is mostly standby, with maybe 2 hours of reading email/HN, messaging, and web browsing. When I want to go mapping, I make sure it's fully charged before I head out; I usually don't want to spend much more than an hour mapping anyway. StreetComplete usage is also a unique situation where, because the battery is drawn more rapidly than other uses, my phone gets hot, which I'm sure doesn't help with the battery life, either; sometimes I take it out of the case to combat this.

**Comment by Helium314** [comment]
In my experience lowering GPS frequency has more than just a slight effect on GPS accuracy, but this might heavily depend on the phone model.
Generally I have the impression that GPS only has a small effect on power consumption, but out of interest I just started a small test that should find some effect: SC with GPS on vs SC with GPS off, just being idle.
[edit: tests ran one hour each, 11% lost with GPS on, 6% lost with GPS off. So I guess maximum possible save (_on my phone!_) is 5%/h]

Switching to the dark theme was a massive improvement for me, the phone definitely gets less warm. Of course this only helps for OLED displays.
Disabling auto-upload also helps a lot: data connection is used much less and CPU load is reduced (if I remember correctly from my v32 tests,  quests are generated on solving _and_ on uploading)

Maybe auto-download could use smaller areas in power save mode, but when moving a lot it might make things worse.

**Comment by mnalis** [comment]
For OLED, special theme where majority of screen is totally black (RGB `#000000`) instead of just "dark" might yield even more improvements, so you might want to play with that.

Disabling auto-rotation of map might help somewhat with CPU/GPU usage too.

But I must say I'm not a big fan of using auto-detection of system settings for this purpose -  better to make if SC-specific preference if your testing shows that it is worthwhile. Because, I usually enable system-wide powersaving mode, and then exclude StreetComplete / OsmAnd from it **exactly** so I would get maximum runtime when mapping (so email, RSS, chat programs, all of android syncers, and all others zillions of programs are disabled from running in background and especially from trying to use data) and thus get all the features from GPS mapping apps.

And `isPowerSaveMode()` looks to be system-wide; I don't know if there exist per-app power-save detection (especially one that isn't tied to specific manufacturer!) which could be used instead.

---
Unreleated for this suggested SC improvement, but related to your use case specifically, if possible I would suggest pre-downloading all area you are going to map while on charger, and then completely disabling android mobile data. 
That helped me a lot in the past on some phones, as it seems something is always slowly nabbing at the data connection if it exist.

Also, does you phone allows to change screen resolution? If it does, lower resolutions do seem to use less power.

**Reply by smichel17** [reply]
I haven't timed myself or done a full 100-to-zero run yet, but from a few short sessions, turning off auto-upload seems to make a really big difference.

FYI, for OLED, while `#000000` does add *some* power savings over just "dark", it's not very much. See https://www.anandtech.com/show/9394/analysing-amoled-power-efficiency, for example. I remember reading a post that compared the power usage of the default Google dark theme to a pure black theme, and the difference was pretty small, to the point where I've stopped using pure black themes unless I decide I like the look better. Unfortunately I can't find the article now.

**Reply by mnalis** [reply]
@smichel17 If I read that `Galaxy S6` graph correctly, it says `0% APL (black)` is about `320 mW`, while `40% APL` at max brightness (quite likely common assumption, or close to it, when outside in the daylight) is at `690 mW`, which is `370 mW` difference. If we assume linearity here, it is about `9.25mW` extra power usage for each `1% APL` increase.

Looking at dark SC screenshot in some example here, background is `#2e2e48` which should be approximately `21.35% APL` (`(46+46+72)/(3*256)`) if I calculate it correctly? So, if your full screen is `SC dark purple grey background` (instead of 100% black `#000000`), that would be `21.35% APL` more, or about `197mW` more power usage at max brightness -- which is huge `62%` power usage increase between "SC dark" and "fully black"!

Of course, background is rarely full screen, but even if empty background is just 1/4 of your screen, switching to pure black from current dark-purple-grey would save you more than `15%` of display power usage, which is not huge, but not negligible either. That is in theory, unless I miscalculated something, or ignored some important information (which I did to simplify; eg. red pixels use less power than green and especially blue ones, for example, but I can't be bothered right now to do better approximation). Anyway, even with most precise calculations (as we know) theory and practice rarely match :smile: .

Anyway, when you are testing, I would suggest to try one run with all quests/maps preloaded, and `mobile data` **completely disabled**. In my tests, that was by far the greatest single win for power saving (followed by reducing screen usage, and its brightness - there are apps that can quickly turn off and on screen, just by waving in front of proximity sensor, so it might help too if quest density is not such that the screen must be on 100% of the time).

**Comment by smichel17** [comment]
I am going to convert this issue to a discussion, since:

- It needs more research about what would be most useful to implement
- I have already volunteered to help with other things and will not get around to this for an indeterminate amount of time  
   and
- It's not even totally clear that reacting to the system setting is desired.

Also I can reply directly to @mnalis' tangent about my phone without disrupting the flow of the main conversation :)

**Comment by HolgerJeromin** [comment]
Perhaps auto upload only every 10 min helps, too.

**Reply by smichel17** [reply]
Yeah, I am going to experiment with what helps battery life, and switching to manual upload is the first thing I will try.

**Comment by Lee-Carre** [comment]
@smichel17

While I'm not at all disagreeing with making StreetComplete more energy-efficient (particularly since it would further enhance my forthcoming suggestion), I have some relevant experience from which to offer suggestions on your surveying expeditions.

**Punchline**: use an external battery pack.

#### Elaboration

I do (on-foot) surveying exclusively (for a few reasons). Normally as a specific / dedicated activity (rather than while happening to do other errands).

To this end, and what with nano-computer batteries being ever more inconvenient to replace (viz non-modular), and current battery chemistry having a quite limited life, I got myself an external battery pack (commonly called a powerbank). The idea being:

* never running low on power during a day, no matter how much I'm using my device, without being tethered to a wall
* offloading wear (from charge cycling; Lithiums are only good for a few hundred full (100% to 0%) cycles; more if discharged less (hence the advice to recharge opportunistically, whenever you can)) from the internal battery to an external; external is ***much*** easier & quicker (and no more costly) to replace (and can be recharged separately) — all externals are compatible (USB standards for the win), rather than having to source a specific battery from a specific manufacturer; thus will also get the benefits of whatever the future brings (chemistry, capacity / density, value (cost per Joule))
* power not being the limiting factor for long outings (surveying or otherwise), and not having to be concerned about having a mains charger & finding a socket
* not paying extortionate prices to recharge at those locker-kiosk things in the likes of cafes
* being able to do more while out (especially surveying); enabling high-power features of various sorts, having more running, being able to do things in ways which aren't the most energy-efficient (but enhance the user-experience, or reduce other resource usage like network transfer or waiting-time), shorter intervals for position & data acquisition (I collect more than just map data, while out), more apps running; don't have to think about any of it
* being able to leave map & surveying apps running in the background, so that they're ready (already have my position & the relevant data loaded) if I want to quickly add or check something opportunistically (e.g. out doing errands and notice something obvious)
* especially for surveying, the limiting factor is *me* (tired, hungry, cold), not equipment; part of that is about making the most of days with suitable weather, without undue faff (finding somewhere to recharge, waiting for charging, carrying excess equipment, mentally context-switching, back-and-forth travel (which doesn't benefit surveying), and whatever other distractions)
* being not limited by power means (to an extent) not being limited by time; so being free to pause surveying to chat with whoever without worry (over time) or having to reconfigure my device; this also means more opportunity to engage in informative conversations about OSM (educating people as to the virtues of non-proprietary, community-maintained maps); in my area it's largely unheard of, with only a few local mappers (at least that contribute *regularly*), so some people are quite curious of what I'm doing when they see me, sometimes that includes the staff in places I'm surveying
* remaining relaxed & leisurely, not in a rush; thus a more pleasant expedition. I often end up discovering something new, while surveying, despite being very local; so it's as much exploration & discovery for me, as it is doing surveying for OSM; being able to pursue my curiosity (where does this (unmapped) path go?) without the nagging feeling of squandering survey-time. When ready, I return to where I was and resume. Another benefit to being leisurely is that if there's a shower, I just wait it out under some cover / shelter, and resume once it stops; because no time pressure.
* this also ties in with safety; gladly waiting at crossings, or taking the safer route, rather than being hasty. There are already enough hazards when surveying.

For context, my longest (single) surveying session was ~12 hours. Most of that was during a very sunny day (so display brightness was at max for most of that), using my device pretty much constantly for the duration, with many apps running. I got lots done, since I was undistracted by concerns like power. Even when external supply was depleted, I still had 100% on the internal, so could just head home.

Without external power, with what I normally have running, the internal (~3Ah) battery depletes at about 1%(age point) per minute. So, at best, that's ~90 minutes, including returning back to base.

To be clear; I leave the two connected, rather than depleting the internal battery merely to recharge it from the external. As far as the device is concerned it's connected to a wall-charger, staying at 100%, while the external (in a pocket) is gradually discharged.

This use-case is different to what's common / intended for external packs. So, be wary of products which assume they'll be recharging a depleted device, and will enter standby mode when current-draw drops low (it thinks the device is now fully charged & has finished its job).

You'll want a decent capacity. My original was a 12Ah, but since that's now dying (I only get ~6 hours out of it, now, but that's after 2+ years of frequent use) I'm gonna be going for a 20+Ah replacement.

If you only do sporadic or brief surveying, then less capacity (given the price difference) may be just fine. Deep discharges take a greater toll, so extra capacity has benefits of longer life & reliability.

Given that non-small capacity will be needed for long sessions, a battery which can be recharged non-slowly is very sensible. As in, a few hours rather than a day.

Brands I'd recommend (have purchased & used):

* [Anker](//www.anker.com/)
* [Hama](//www.hama.com/) (not as good as Anker, but still decent & above average)

Anker (battery pack) products usually have a trickle-charge mode / feature, which disables auto-standby (misfeature) when draw is low. Given the gotcha I mentioned earlier and just needing to keep the device's internal battery topped-up (especially if you're using only StreetComplete).

Brands which have been recommended to me (but I've not tried, only researched, though seemed promising candidates)

* [SandBerg](//www.sandberg.world/)

You'll also want non-fragile products, since they won't simply be sitting on a desk. To that end I've been advised (by a retailer) to **avoid** Vibe products, which he described as liable to break with any drop.

To that end, I'd also recommend a case (Hama sell small accessory bags); both to protect the pack (if damaged, the case is trivial & cheap to replace) while in-pocket when you're out walking along, and when not actively using your device to keep them together, zipped up, in a single bundle (easier handling, no cable snags).

Depending on your circumstances, you may also consider an adaptor to charge from your car's cigar lighter (both Anker & Hama sell good designs of such adaptors; bought & used them). Thus, can recharge when travelling between survey sites.

Don't bother with silliness like inductive (‘wireless’) charging (it's significantly less efficient than by-cable), or ‘magnetic’ ones. No gimmicks, just a quality battery pack.

Depending where you are, and if such things matter to you:

* Anker is based in Yankland
* both Hama & Sandberg are (northern) Europe (Germany & Denmark, respectively)

For recharging; I'd strongly recommend USB PowerDelivery.

* Hama has some, but they tend to be of a lower rate (30W input on a 25Ah pack, or 18W on a 12Ah).
* Anker offers many USB PD products, and battery packs (of large capacity) up to 60W (input & output). Thus a 25Ah pack can be recharged in 3 hours (rather than ~24+). Anker's products for charging USB from the mains are excellent.
* Sandberg appears to have decent-capacity packs which the specs say can handle 100W (the maximum of USB PD).

If you don't survey frequently (multiple consecutive days) then recharge rate may not be of concern (and slower is probably better for the cells, anyway). In that case, especially for storage, keep it between 20% & 80% charged (oddities of Lithium battery chemistry, other chemistries are different again). Definitely don't leave it with power supplied for extended periods (viz so that it's always ready at 100%, despite the convenience of this).

Packs that *can* be recharged quickly don't have to be; just connect a lower-output supply. It's then an option for when time is limited, rather than it *always* being slow.

I'd also recommend investing in decent cables; nylon-braided, metal connectors. They shouldn't be expensive. I can't make recommendations of brands since my local retailer sells unbranded ones 🙂. Though, I suppose PVC is more water-proof, if not as hard-wearing as nylon.

Other tips (for extended surveying) include:

* keep the battery pack (and any batteries, actually) at moderate temperatures; in an external pocket during summer, to allow air flow, and in cold weather somewhere (maybe an inside chest pocket) where it won't become cold (preferably zipped up, with only a cable of sufficient length protruding)
* especially in cold weather, unless you're trekking along, it's easy to become cold (especially after sundown), so ample clothing, especially for your hands (cold fingers don't function well, especially when needing fine motor control). Though this has been more the case with using Vespucci (and doing far more than just re-tagging), than with SC.
* for being able to take photos, you'll want a non-short cable, so that you can freely position, orient / rotate, and angle the camera without yanking on said cable. A 1m cable would probably do nicely; enough slack without needing to coil it.

If you do end up getting into more intensive surveying, then to recharge everything when back at home, an Anker charging hub would be good. The non-basic ones have multiple high output ports (including USB PD), to connect packs & devices simultaneously, ready for next time.

Hope this helps.

**Reply by westnordost** [reply]
Interesting insight. I am also concerned about the trend that every device has a non-removable Li-Ion inside and the e-waste it produces. For this reason, I got myself a power bank that is at the same time a charger for standard Li-Ion 18650 cells which can be found in (old) notebooks, batteries for e-bikes, many power banks etc. and is used in (high-end) flashlights, e-cigarettes and more. This way, I can recycle (or, well, use to the end) cells of old notebooks that otherwise became too feeble. If one cell reached the end of its lifespan, I can simply replace it with another standard cell.

![20211014_191804](https://user-images.githubusercontent.com/4661658/137366602-6ef92264-de1f-4fba-8f72-3c45cca0c7d3.jpg)
![20211014_191811](https://user-images.githubusercontent.com/4661658/137366607-b2d033b2-48a3-4208-845e-e6c75d0f6dbc.jpg)

The particular cells in the power bank shown should now almost be 10 years old.

**Reply by Lee-Carre** [reply]
> Interesting insight.

Thanks. Been meaning to do a write-up (as\|for) an [OSM Diary](//www.openstreetmap.org/user/Lee%20Carré/diary) entry, since other surveyors may benefit.

>  I am also concerned about the trend that every device has a non-removable Li-Ion inside and the e-waste it produces. For this reason, I got myself a power bank that is at the same time a charger for standard Li-Ion 18650 cells which can be found in (old) notebooks, batteries for e-bikes, many power banks etc. and is used in (high-end) flashlights, e-cigarettes and more. This way, I can recycle (or, well, use to the end) cells of old notebooks that otherwise became too feeble. If one cell reached the end of its lifespan, I can simply replace it with another standard cell.

Ah! So, such things *do* exist! 😃 Excellent; thankyou for this pointer (& enough terms / keywords (especially the cell-type number) to begin (re)searching).

I'd wondered about this arrangement for a long time; standardised cells (both dimensions / interface & electrical, like AA or C are) which can be recharged in a reusable charger.

I'm old enough to recall when this was normal practise for NiCd / NiMH cells. Fell out of fashion, though.

Especially that Li AA & AAA cells are available (having purchased & used them, before).

Always seemed daft that one has to dispose of the entire product (case, electronics, and all), when it's only the cells that are inoperative.

Maybe it's a bit of the-razor-blades-model of marketing.

Makes me think of Right to Repair (& [Lewis Rossmann](//www.youtube.com/channel/UCl2mFZoRqjw_ELax4Yisf6w)).

I think part of the trend is that many people (especially where I am) treat nano-computers as fashion accessories (to the point that I coined the term Pink Phone Syndrome (the form & colour is valued much more than the functionality of what's inside)), and replace them often (every 12-18 months). Thus, if the device is replaced before the internal battery fails, why go to the expense of designing & manufacturing a modular battery in the first place?

Not that I'm denying that devices are designed to be anti-repair. Old enough, and have seen enough, to be too cynical for that.

Markets respond to demand, though. [Moen's Law of Bicycles](http://linuxmafia.com/~rick/lexicon.html#moenslaw-bicycles) (“Good customers make for good products” (that aren't artificially overpriced)).

Unfortunately, being more than a (passive) consumer is the exception rather than the rule.

More generally, even if everyone used modular cells, that's still a lot of lithium to be mined. I read that graphene shows promising potential in battery applications; I'm not sure that it'll *replace* lithium, but it seems to enhance the electrical properties (so less lithium for the same energy storage?).

Frustratingly, being on an island (and lithium regarded as a hazardous material; lots of regulation & paperwork for shipping), even getting a decent battery pack (let alone a specific replacement (internal) battery, say for a laptop) is something of a challenge (without resorting to such evils as [Amazon](//www.stallman.org/amazon.html)). So, I seriously doubt that I'd be able to source a supply of bare cells.

However, I'll certainly keep this suggestion / recommendation in mind.

**Reply by westnordost** [reply]
Regarding Li-Ion cells, you must be cautious. You should buy from trusted sources (e.g. brands) even if they are more expensive or "recycle" old Li-ons by carefully prying open old battery packs from brands that contain these (for laptops etc).

You need to know another important distinction: There are cells with a protection circuit (against deep discharge, overcharge, short circuit etc) and those without. Read more here: https://www.fenix-store.com/blog/what-is-the-difference-between-protected-and-unprotected-18650-batteries/
Usually cells in battery packs do not have a protection circuit because there is already a protection circuit in the pack itself.

You can buy 18650 also in any shop that sells e-cigarettes, maybe there is one on Jersey.

**Reply by westnordost** [reply]
The tomo powerbanks/chargers have reverse-polarity, deep-discharge etc protection also built in, so protected cells are not necessary by the way. They come **without** the cells, so there shouldn't be a problem importing these.
As said, you get new 18650 cells at e.g. e-cigarette store but also stores that sell high-end flashlights.

**Comment by smichel17** [comment]
> Punchline: use an external battery pack.

@Lee-Carre I've tried that. It… works, but it's awkward, and if the cable doesn't fit snugly it's apt to rapidly connect & disconnect while walking. I'd much rather own a phone with a replaceable battery (fond memories of my old **[Galaxy S4](https://www.youtube.com/watch?v=5sDpifWuJ9Q&t=15s)**) and carry around extras to swap as needed.

I don't typically go on extended mapping sessions; 1h is fairly common but longer than was rare even before my battery degraded as much as it is now. Also, turning off auto-upload helped a massive amount. I'm not sure *exactly* how much (especially since I know my phone indicator lies sometimes) but I got at least half an hour more time. I was out for an hour and 45 minutes a few days ago and got back with some charge left! (30%, I think?). I'll see if we can make the dark theme usable during the day (#3378), and probably be content with that.

**Reply by mnalis** [reply]
I've had some luck in the past with spiral cables (like on old telephone handsets, eg. like [this](http://web.archive.org/web/20250304223310/https://i.ebayimg.com/images/g/DuYAAOSw-LZl-qsQ/s-l1600.jpg)), and having a battery bank in a jacket pocket, with spiral cable stretching as needed to phone in a hand. It had a distinctive [Blake's 7 liberator gun](http://d2f0ora2gkri0g.cloudfront.net/bkpam2152728_libby1small.jpg) look, though :smile:

**Comment by matkoniecz** [comment]
I am worried that this would result in people confused why SC is behaving in a substandard way and not connecting it to some system setting - there is already repeated confusion with sound setting following system setting.

Like for #3556

**Reply by mnalis** [reply]
perhaps the toggle could be added in the same way as for `light/dark theme` for example (ie. _"Data saver: on / off / follow system setting"_) ? That should avoid confusion issues, but adds an extra preference (which is not ideal for SC, as known).

**Reply by smichel17** [reply]
Some other app I use, don't remember which, has a "Notification settings" item which actually just opens up the system notification settings for the app. I'm not sure if such a thing is possible for data saver (or the sound setting), but it's an interesting approach.

**Reply by Helium314** [reply]
> don't remember which

Signal does that.
Opening other settings screens should be no problem, except when vendor has decided to arrange settings differently...

**Comment by mnalis** [comment]
perhaps reducing number of screen refreshes should help with battery usage; see https://github.com/streetcomplete/StreetComplete/issues/966#issuecomment-1065884412
